contributors/design-proposals/api-machinery/customresources-versioning.md

```diff
@@ -0,0 +1,177 @@
+  // flagged as Storage.
+  Storage Boolean
+  //
+  Conversion CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
+  Declarative map[string]CustomResourceDeclarativeConversion
+}
+
+type CustomResourceDeclarativeConversion {
+  // This type will have collection of conversion operations. Only one of them should be set.
+  Rename RenameConversion
+}
+
+type RenameConversion {
+  From JsonPath
+  To JsonPath
+}
+
+type JsonPath string
+```
+
+# **External Conversion/Webhooks**
+
+Webhooks gives flexibility to CRD author to convert from the specific version to any version. As the conversion needs to be reversible, here we only list versions this webhook converts to/from. 
+
+```golang
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+  //
+  Conversion CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
+  Declarative map[string]CustomResourceDeclarativeConversion
+  External CustomResourceConversionWebhook
+}
+
+type CustomResourceExternalConversion {
+  // List of versions this external conversion supports.
+  To []string
+  Webhook ConversionWebhook
+}
+
+type ConversionWebhook struct {
+  // ClientConfig defines how to communicate with the hook.
+  ClientConfig WebhookClientConfig `json:"clientConfig" protobuf:"bytes,2,opt,name=clientConfig"`
+}
+```
+
+This API design allows multiple webhooks but that should be prevented in validation. Also there is no need to redefine webhook for all versions. For example if on "v1" version we specified a webhook to convert it to “v2”, we do not need to add that webhook configuration to “v2” as the conversion is defined in both ways.
+
+**Note: ** Webhooks will, intentionally, be not included in the first version of the feature to make sure authors waiting for this feature is trying the declarative approach first. The Webhook approach is a fallback for all cases that declarative cannot support.
+
+# **Validation**
+
+The API structure above allows some invalid states in order to keep a clean easy API. There need to be validation to catch undesire states. The validations would be
+
+* Multiple webhooks are not allowed for a CRD. This will help us improve the performance by keeping the connection open to the webhook if the user defined a chain of version conversions.
+* There should be no loop in the conversion graph. This includes v1->v2->v1 or v1->v2->v3->v1. The former is implicit in the assumption that webhook conversion is round-trippable. The second one allows two way to convert v1 to v3 which is not desired.
+* Rename operation can be limited at first to not change the structure despite the fact that `To` field is also a json path.
```

Comment:
See suggestion above: we only want to rename fields in objects, no moves, no structure changes.